Partially fix "Dummy" actually using ALSA

M374LX opened the pull request:
Partial fix for #2349.

Even without this fix, an error opening the desired device seems to cause the setup dialog to be shown instead of simply falling back to Dummy. Also, the following message is shown on the console:
```
*** Error in `lmms': free(): invalid size: 0x00000000020169b0 ***
```

tresf commented:
Looks great.  Not sure about the re-showing of the dialog problem, but I suspect that it is an existing problem.  Clean shuffling of operations here to honor an explicit `Dummy` selection.

The only problem that I have with this approach in general is that the Dummy state is undesired almost always, so we should have a fallback state AND a dummy state so that dummy is explicitly chosen when it is explicitly selected.  As mentioned in the bug report, I don't think this is possible, so we may decide to keep the counter-intuitive behavior in lieu of better sane defaults.

M374LX commented:
In `MainWindow::finalize`, we have:
```C++
// look whether mixer could use a audio-interface beside AudioDummy
else if( Engine::mixer()->audioDevName() == AudioDummy::name() )
{
	// no, so we offer setup-dialog with audio-settings...
	SetupDialog sd( SetupDialog::AudioSettings );
	sd.exec();
}
```

As we can see, the settings dialog will be shown every time Dummy is selected by the program (either explicitly or as a fallback). A simple solution is to set a boolean variable to `true` inside `Mixer` when the fallback happens and read such value by a getter method.

M374LX commented:
In `AudioDummy::run()`, we have:
```C++
const surroundSampleFrame* b = mixer()->nextBuffer();
if( !b )
{
	break;
}
```

The cause of the error message when using DummyAudio and closing the program seems to be `Mixer::nextBuffer()` returning a wrong value.

tresf commented:
Somehow I missed the additional work you did on this.  Thanks.  Once someone has time to test, we'll need you to squash the commits and then we can merge the PR.

tresf commented:
Again... we can merge this once we get some valid testing and the commits are merged.

M374LX commented:
Except for the message on the console, everything went fine on my own tests some time ago, but I am going to do new tests today.

I think the message on the console is a topic for a new issue.

tresf commented:
> I am going to do new tests today.

Thanks for the prompt reply.  I'll tag a few testers... @Umcaruje @Wallacoloo.

@musikBear if you had a dev environment setup, this would be a great one for you to test.

tresf commented:
@M374LX why did you close this?

M374LX commented:
> @M374LX why did you close this?

Having trouble with Git, as usual.

tresf commented:
:+1:

M374LX commented:
Finally, all changes in a single commit.

Here are my tests:
1. LMMS started with ALSA selected; then I changed to Dummy.
2. LMMS started with Dummy selected; it worked fine with no sound, as expected; then I changed to OSS; the same error message was shown on closing the program, but this is a topic for a new issue, as said earlier.
3. OSS failed on my system; then the settings dialog was show while the audio device fell back to Dummy; I changed the device to ALSA; same results with Dummy.
4. With ALSA selected on the previous test, ALSA worked fine again.

tresf commented:
@M374LX did you try to move/remove the `~/.lmmsrc.xml` file too to confirm that behaves normally?

M374LX commented:
> @M374LX did you try to move/remove the ~/.lmmsrc.xml file too to confirm that behaves normally?

By removing the file, the settings dialog is shown, but it looks like the program uses the first audio device available (ALSA in my system). Knowing that the audio device is initialized before the settings dialog is shown, this is perfectly logical. Again, this is a topic for a new issue.

tresf commented:
> ALSA in my system


So SDL isn't available I take it? It should be first chosen via https://github.com/LMMS/lmms/commit/1bb276b3654e286b1a4be7082824612f94a05ed1 if it is available.

M374LX commented:
> So SDL isn't available I take it? It should be first chosen via 1bb276b if it is available.

Somehow, SDL seems to have been not detected on my system during the build process. It is not even available on the settings dialog.

tresf commented:
Fair enough.  Thanks for the fix, testing, squashing.  Top notch.  Merging.

musikBear commented:
@tresf sorry, i dont have it- My old box dont have room for a linux partition 
I do have puppy linux on a rescue-disk, though, but absolutely NO idea how to neither get or use a linux distro of lmms, and does not know if a puppy will run lmms at all
I think there is too many unknowns in the equation, to actually make a valid result for anything.. sorry

tresf commented:
It's tested and merged now, no worries.  Want me to ship you a hard drive?

musikBear commented:
> Want me to ship you a hard drive?
@tresf 
:P doubling as santa! 
 I expect changes i near future. :)

tresf commented:
> doubling as santa! 

Just trying to get some help, that's all. :+1:
